lotman/lotman_test.go

```diff
@@ -363,3 +394,602 @@ func TestDeleteLotsRec(t *testing.T) {
+			if len(tc.errorStrings) > 0 {
+				require.Error(t, err)
+				for _, errStr := range tc.errorStrings {
+					require.Contains(t, err.Error(), errStr)
+				}
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+// Make sure we handle various suffixes and non-suffixed percentages correctly
+func TestConvertWatermarkToBytes(t *testing.T) {
+	type testCase struct {
+		Name          string
+		Watermark     string
+		Expected      uint64
+		ErrorExpected bool
+	}
+
+	totDisk := uint64(1000000000000) // 1TB
+	testCases := []testCase{
+		{
+			Name:          "Valid 'k' suffix",
+			Watermark:     "100k",
+			Expected:      uint64(100000), // 100KB
+			ErrorExpected: false,
+		},
+		{
+			Name:          "Valid 'm' suffix",
+			Watermark:     "100m",
+			Expected:      uint64(100000000), // 100MB
+			ErrorExpected: false,
+		},
+		{
+			Name:          "Valid 'g' suffix",
+			Watermark:     "100g",
+			Expected:      uint64(100000000000), // 100GB
+			ErrorExpected: false,
+		},
+		{
+			Name:          "Valid 't' suffix",
+			Watermark:     "100t",
+			Expected:      uint64(100000000000000), // 100TB
+			ErrorExpected: false,
+		},
+		{
+			Name:          "No suffix is percentage",
+			Watermark:     "50",
+			Expected:      uint64(500000000000), // 500GB
+			ErrorExpected: false,
+		},
+		{
+			Name:          "Invalid suffix",
+			Watermark:     "100z",
+			Expected:      uint64(0),
+			ErrorExpected: true,
+		},
+		{
+			Name:          "Invalid value",
+			Watermark:     "foo",
+			Expected:      uint64(0),
+			ErrorExpected: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.Name, func(t *testing.T) {
+			result, err := convertWatermarkToBytes(tc.Watermark, totDisk)
+			if tc.ErrorExpected {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				require.Equal(t, fmt.Sprintf("%d", tc.Expected), fmt.Sprintf("%d", result))
+			}
+		})
+	}
+}
+
+// If so configured, we'll divide unallocated space between lot's dedicated GB and
+// opportunistict GB values. This test ensures the calculations are correct and that
+// hardcoded configuration isn't modified.
+// I don't test for errors here because the internal functions capable of generating
+// errors are tested elsewhere (e.g. convertWatermarkToBytes)
+func TestDivideRemainingSpace(t *testing.T) {
+	server_utils.ResetTestState()
+	defer server_utils.ResetTestState()
+	dedGB := float64(10.0)
+	oppGB := float64(1.5)
+	lotMap := map[string]Lot{
+		"lot1": {
+			LotName: "lot1",
+			MPA: &MPA{
+				DedicatedGB:     &dedGB,
+				OpportunisticGB: &oppGB,
+			},
+		},
+		"lot2": {
+			LotName: "lot2",
+			MPA:     &MPA{},
+		},
+		"lot3": {
+			LotName: "lot3",
+			MPA: &MPA{
+				DedicatedGB: &dedGB,
+			},
+		},
+		"lot4": {
+			LotName: "lot4",
+			MPA: &MPA{
+				OpportunisticGB: &oppGB, // hardcoded values should be respected
+			},
+		},
+	}
+
+	totalDiskSpaceB := uint64(30000000000) // 30GB
+	viper.Set("Cache.HighWaterMark", "25g")
```

Comment:
What happens when the Cache.HighWaterMark is poorly set and lower than the combined total of the dedicatedGB?

Reply:
This was a good point, and I'm glad I poked at the code here. Because `hwm` is represented as a uint64, there was an uncaught underflow that caused it to wrap and had unexpected behavior. Thanks for pointing it out!